tests/std/tests/VSO_0000000_regex_use/test.cpp

```diff
@@ -546,6 +546,17 @@ void test_VSO_226914_word_boundaries() {
     aWordAny.should_search_fail("aa", match_not_bow | match_not_eow);
 }
 
+void test_gh_731() {
+    std::vector<std::string> groups{"AAA", "BBB", ""};
+    g_regexTester.should_capture_groups("AAA BBB", R"((A+)\s*(B+)?\s*B*(B*)?)", groups);
+
+    groups.assign({"AAA", "BBB"});
+    g_regexTester.should_capture_groups("AAA BBB", R"((A+)\s*(B+)?\s*B*)", groups);
+
+    groups.assign({"c"});
+    g_regexTester.should_capture_groups("aabaac", "(aa|aabaac|ba|b|c)*", groups, std::regex_constants::extended);

```

Comment:
I think we need additional test coverage to exercise `_Better_match`, especially scenarios where ECMAScript and POSIX Extended accept the same syntax but have different ideas about what should be captured. I think we should also manually verify agreement with Boost/libstdc++/libc++, or understand any remaining divergence. (e.g. I observed MSVC, libstdc++, and libc++ getting 3 sets of answers for the subject `"WXconcatenateYZ"` with the pattern `".*(cat|concatenate)"` https://godbolt.org/z/bza5nrYcM , woof :dog:).